app/src/main/java/chat/rocket/android/push/PushManager.kt

```diff
@@ -0,0 +1,604 @@
+package chat.rocket.android.push
+
+import android.annotation.SuppressLint
+import android.app.Notification
+import android.app.NotificationChannel
+import android.app.NotificationManager
+import android.app.PendingIntent
+import android.content.BroadcastReceiver
+import android.content.Context
+import android.content.Intent
+import android.media.RingtoneManager
+import android.os.Build
+import android.os.Bundle
+import android.support.annotation.RequiresApi
+import android.support.v4.app.NotificationCompat
+import android.support.v4.app.NotificationManagerCompat
+import android.support.v4.app.RemoteInput
+import android.text.Html
+import android.text.Spanned
+import android.util.Log
+import chat.rocket.android.BuildConfig
+import chat.rocket.android.R
+import chat.rocket.android.chatrooms.ui.MainActivity
+import org.json.JSONObject
+import java.io.Serializable
+import java.util.*
+import java.util.concurrent.atomic.AtomicInteger
+import kotlin.collections.HashMap
+
+typealias TupleGroupIdMessageCount = Pair<Int, AtomicInteger>
+
+/**
+ * Refer to: https://github.com/RocketChat/Rocket.Chat.Android/blob/9e846b7fde8fe0c74b9e0117c37ce49293308db5/app/src/main/java/chat/rocket/android/push/PushManager.kt
+ * for old source code.
+ */
+object PushManager {
+    const val EXTRA_NOT_ID = "chat.rocket.android.EXTRA_NOT_ID"
+    const val EXTRA_HOSTNAME = "chat.rocket.android.EXTRA_HOSTNAME"
+    const val EXTRA_PUSH_MESSAGE = "chat.rocket.android.EXTRA_PUSH_MESSAGE"
+    const val EXTRA_ROOM_ID = "chat.rocket.android.EXTRA_ROOM_ID"
+    private const val REPLY_LABEL = "REPLY"
+    private const val REMOTE_INPUT_REPLY = "REMOTE_INPUT_REPLY"
+
+    // Notifications received from the same server are grouped in a single bundled notification.
+    // This map associates a host to a group id.
+    private val groupMap = HashMap<String, TupleGroupIdMessageCount>()
+
+    // Map a hostname to a list of push messages that pertain to it.
+    private val hostToPushMessageList = HashMap<String, MutableList<PushMessage>>()
+    private val randomizer = Random()
+
+    /**
+     * Handles a receiving push by creating and displaying an appropriate notification based
+     * on the *data* param bundle received.
+     */
+    @Synchronized
+    fun handle(context: Context, data: Bundle) {
+        val appContext = context.applicationContext
+        val message = data["message"] as String?
+        val image = data["image"] as String?
+        val ejson = data["ejson"] as String?
+        val notId = data["notId"] as String? ?: randomizer.nextInt().toString()
+        val style = data["style"] as String?
+        val summaryText = data["summaryText"] as String?
+        val count = data["count"] as String?
+        val title = data["title"] as String?
+
+        if (ejson == null || message == null || title == null) {
+            return
+        }
+
+        val lastPushMessage = PushMessage(title, message, image, ejson, count, notId, summaryText, style)
```

Comment:
why lastPushMessage and not just pushMessage? With that name seems that you are dealing with an old/previous push message